### ISSUE TITLE
Compilation errors when building from source

### ISSUE THREAD

**Issue by lczech** [popgenmethods/smcpp#160]
When building from source, I run into compilation errors, both when trying
```
pip install git+https://github.com/popgenmethods/smcpp
```
as well as cloning followed by
```
sudo python setup.py install
```

Both of them fail:
```
src/conditioned_sfs.cpp:108:27: error: ‘n’ was not declared in this scope
         assert(csfs[i](2, n) == 0.);
                           ^
```

Hence, in the cloned repository, I commented out that line, and got further errors:
```
include/marginalize_key.h: In static member function ‘static std::map<block_key, double> marginalize_key<P>::run(const Eigen::MatrixBase<OtherDerived>&, const Eigen::MatrixBase<OtherDerived>&, const Eigen::MatrixBase<ScaleDerived>&)’:
include/marginalize_key.h:61:12: error: there are no arguments to ‘a’ that depend on a template parameter, so a declaration of ‘a’ must be available [-fpermissive]
     assert(a(0) >= 0);
            ^
```

After commenting out that line as well, I can compile. 

Update: First, after compiling, I still had issues, because I was using the [wrong Python version](https://github.com/popgenmethods/smcpp/issues/161#issue-640846262), but after fixing that, the above error messages did not occur any more (in a fresh clone of the repository, with Python 3.7 activated). Instead, there were a ton of warnings, but no errors related to compilation. Hooray. Still, the above errors might be worth fixing.

**Comment by terhorst** [comment]
Thanks for all the bug reports. I am not sure why you encountered that error. According to `git blame`, that line has been there for over three years, during which time the code has been compiled many times by myself and others. Feel free to re-open if you are able to determine the source of the error.

**Comment by lczech** [comment]
Hm, I tried to figure out what exactly happened there. As far as I can tell, when the above compilation errors occurred, my `python` linked to some Python 2 version. Then, I changed `python` to refer to Python 3.7 instead, and the error disappeared. 

I have no idea why the python version in the environment affects the cpp compilation. Are there any debug flags that depend on that? As the variables that the compilation error complains about are not declared in the respective scope, but were only used within assertions, my guess is that the `NDEBUG` flag was set when using Python 3, but not when using Python 2... No idea how that happens, but you know your build setup better, so maybe that helps to figure out what is going on ;-)
